### ISSUE TITLE
Feature request: Ability to register MacOS services

### ISSUE THREAD

**Issue by ccnokes** [electron/electron#8394]
I would like to be able to register a "service" (MacOS only I think) so that I can, for example, add custom context menu items to all files on the file system, like so:

<img width="763" alt="screen shot 2017-01-11 at 10 40 11 pm" src="https://cloud.githubusercontent.com/assets/1610343/21878361/f9c0a4c8-d84e-11e6-9ec6-3bf4b2ad9cb3.png">

Apple overview: https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/SysServices/Articles/overview.html#//apple_ref/doc/uid/20000850-BCIDHJJA

I'm not sure what this would take technically but it would be cool and something that native apps have that Electron doesn't yet.

**Comment by TimNZ** [comment]
Interested in Electron team thoughts on this as I want to understand how decision is made that something is core to mandate vs should be an addon.  

This seems to sit in the middle.

I expect the short term solution would be to build a native nodejs module.

**Comment by kevinsawicki** [comment]
> Interested in Electron team thoughts on this as I want to understand how decision is made that something is core to mandate vs should be an addon.

Yeah, briefly reading https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/SysServices/Articles/providing.html#//apple_ref/doc/uid/20000853-BABDICBI makes it seems like this could be something completely managed via a native module plus some `.plist` changes.

**Comment by ccnokes** [comment]
@kevinsawicki If you guys have written up some criteria on what goes in core and what doesn't, I'll be sure to read that before opening a feature request issue. I haven't seen anything like that though...

**Comment by kevinsawicki** [comment]
>  If you guys have written up some criteria on what goes in core and what doesn't, I'll be sure to read that before opening a feature request issue. I haven't seen anything like that though...

We don't have that currently, it is totally a case by case basis. But I think it was great that you opened this issue up here, it is good to have this discussion since this seems like a valuable feature that many apps would want to take advantage of.

Some APIs can't be done as native node modules because Electron does not provide the objects/hooks needed. Other times Chrome already has great APIs to do things like this and so it is straightforward to add public APIs to Electron that call directly into these Chrome APIs.

As a next step, perhaps someone could spike out a node module that registers a service based on https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/SysServices/Articles/providing.html#//apple_ref/doc/uid/20000853-BABDICBI and see if it is possible or if Electron would need provide something to enable it.

**Comment by willnewii** [comment]
i need same ability  :-)
